Parse -s changes using the json parser where possible, allowing human-readable names in -s lists

kripken opened the pull request:
We have nice syntax for lists, `-s LIST=[foo,bar]` which works well in many cases. But sometimes the list contains items that contain `,`, for example `-s LIST=["foo(x, y)", "bar"]`. To handle that properly, try to parse things that look like json using the json parser; if it fails go back to the old code.

This will help https://github.com/emscripten-core/emscripten/issues/9128 where the lists there actually do have `,`s (after the binaryen part lands as well, a test for this will be added).

quantum5 commented on emcc.py:
```diff
@@ -3493,7 +3493,20 @@ def parse_string_list(text):
+
   if text[0] == '[':
+    # proper json can like ["a", "b"] can be simply handled by the python json parser
+    if text[1] == '"':
```
Do we need to check this? The JSON would fail to parse `[a, b]`, and would fall back to `parse_string_list`.

kripken replied:
True, but it may be faster to avoid even entering the JSON code unnecessarily, and I also think it makes the code clearer as to what we can parse in this path and what we can't.

kripken replied:
Fair enough, I refactored it now following your suggestion and @quantum5's to avoid that check, which also lets us handle more json (including whitespace after the `[`).

kripken commented:
As the Binaryen side here has landed, I added the new tests here, so this PR is fully tested. Summary for squash:

===
We have nice syntax for lists, -s LIST=[foo,bar] which works well in many cases. But sometimes the list contains items that contain ,, for example -s LIST=["foo(x, y)", "bar"]. To handle that properly, try to parse things that look like json using the json parser; if it fails go back to the old code.

Together with https://github.com/WebAssembly/binaryen/pull/2275 this fixes #9128 where the lists there actually do have ,s in them.

Add testing for Asyncify lists with such names, and also added extra warnings for when escaping is confusing, and a test for that as well.
===

kripken commented:
Added a fix for newlines in the parsed string, which can happen in response files. This implements @quantum5 's suggestion to use the json parser without checking for the second character being `"`.

kripken commented on emcc.py:
```diff
@@ -3493,7 +3506,20 @@ def parse_string_list(text):
+      return parse_string_list(text)
+
   if text[0] == '[':
+    # proper json can like ["a", "b"] can be simply handled by the python json parser
```
Thanks! In the last refactor I removed this code anyhow.

kripken commented:
The binaryen change actually did not land yet, due to https://github.com/WebAssembly/binaryen/pull/2271#issuecomment-517446827 - so this can't pass tests yet. I pushed some unrelated fixes meanwhile (I forgot to disable some new tests on fastcomp).

sbc100 commented:
I think this is pushing the the limits of what a command line parser should reasonable be expected to handle.   Whats wrong with simply requiring that all symbols be C symbols (i.e. name mangled)?

quantum5 commented:
I think the rationale is that in the WASM binary, we always use the demangled names.

kripken commented:
@sbc100 The wasm backend emits human-readable names in the Names section. So those are the names we have to use in a list like this. The reason this doesn't come up in say EXPORTED_FUNCTIONS is that we pass that list to wasm-ld, which I guess handles the mangling internally? For the asyncify lists, we look in the Names section of the binary wasm-ld emitted.

If the wasm backend had a flag to emit the mangled names instead, that could help. However, the human-readable names are kind of nice for reading stack traces and profiling data. So there isn't an obvious solution here.

sbc100 commented:
I see.

As an aside I'm working on a change for a simplified `-s` list format that looks more like:

`-s SETTING=foo,bar`

and a simplified list file where `-s SETTING=@file.txt` where file would simply been list of items one per line.

This would make both the command line and the @listfile files much easier to generate and read, but I think it would force anyone wanting to put a comma inside one of the list items like this to use the @listfile method.   I don't anyone wants to type full C++ method signatures into the command line anyway do they?

kripken commented:
@sbc100 I'd be ok with not allowing full C++ names on the commandline, if it works with a response file (I think it already would after this PR), and if there's a clear error pointing people to use a response file when necessary.
